src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs

```diff
@@ -1,52 +1,91 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System;
+using System.Collections.Immutable;
 using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Formatting
 {
     public static class FormattingOptions
     {
-        public static PerLanguageOption<bool> UseTabs { get; } =
-            new PerLanguageOption<bool>(nameof(FormattingOptions), nameof(UseTabs), defaultValue: false,
-                storageLocations: new EditorConfigStorageLocation<bool>("indent_style", s => s == "tab"));
+        private static readonly ImmutableArray<IOption>.Builder s_allOptionsBuilder = ImmutableArray.CreateBuilder<IOption>();
+
+        static FormattingOptions()
+        {
+            AllOptions = s_allOptionsBuilder.ToImmutable();
```

Comment:
i'm very confused what's happening here.  it looks like this just gets initialized to an empty array...

Reply:
Static field initializers are executed prior to static constructor, and each field initializer is adding the created option to `s_allOptionsBuilder`. We basically need to create a collection of static readonly field values, without explicitly having separate code that lists each option, and this was the only approach that came to my mind. Let me know if you have a better approach in mind. Otherwise, I can add a comment here for clarity.

---
In reply to: [219243277](https://github.com/dotnet/roslyn/pull/30043#discussion_r219243277) [](ancestors = 219243277)

Reply:
At the very least, it would be nice if the static constructor came last :)  and it would be good to have an explanation here of how this depends on the fields being initialized first.  

An alternate approach would also just be to dynamically craete the AllOptions array on demand.  There's no strong need for it to have to be statically initialized up front.

Reply:
I will move the .cctor at the end and add comments.

> An alternate approach would also just be to dynamically craete the AllOptions array on demand.

This initialization will still need to explicitly list/add all the options? I wanted to avoid users adding more options in future and forgetting to updating the AllOptions initialization.

---
In reply to: [219301049](https://github.com/dotnet/roslyn/pull/30043#discussion_r219301049) [](ancestors = 219301049)

Reply:
> This initialization will still need to explicitly list/add all the options? I wanted to avoid users adding more options in future and forgetting to updating the AllOptions initialization.

Right.  I would still keep the approach you ahve wrt CreateOption adding to the list.  The only change would be that calling AllOptions would then just call .ToImmutable at that point.

THere it would be more obvious (to me at least) that the code of a property runs after all the fields are initialized.  as opposed to a static construction, which (for me) was more confusing since it wasn't super clear it ran after fields.

Note: i don't feel strongly about this :)   I'm happy with just the static-constructor at the end :)